includes/cron/class-urlslab-serp-cron.php

```diff
@@ -0,0 +1,250 @@
+		$request->setLocale( $query->get_lang() );
+		$request->setCountry( $query->get_country() );
+		$request->setAllResults( true );
+		$request->setNotOlderThan( $widget->get_option( Urlslab_Serp::SETTING_NAME_SYNC_FREQ ) );
+		$has_monitored_domain = false;
+		$urls                 = array();
+		$domains              = array();
+		$positions            = array();
+
+		try {
+			$serp_response = $this->serp_client->search( $request );
+			$organic       = $serp_response->getOrganicResults();
+
+			if ( empty( $organic ) ) {
+				$query->set_status( Urlslab_Serp_Query_Row::STATUS_NOT_PROCESSED );
+				$query->update();
+
+				return false;
+			} else {
+				foreach ( $organic as $organic_result ) {
+					$url_obj = new Urlslab_Url( $organic_result->link, true );
+					if ( isset( $this->domains[ $url_obj->get_domain_id() ] ) && $organic_result->position <= $widget->get_option( Urlslab_Serp::SETTING_NAME_IMPORT_RELATED_QUERIES_POSITION ) ) {
+						$has_monitored_domain = true;
+					}
+
+					if ( 10 >= $organic_result->position || $this->domains[ $url_obj->get_domain_id() ] ) {
+						$url    = new Urlslab_Serp_Url_Row(
+							array(
+								'url_name'        => $organic_result->link,
+								'url_title'       => $organic_result->title,
+								'url_description' => $organic_result->snippet,
+								'url_id'          => $url_obj->get_url_id(),
+								'domain_id'       => $url_obj->get_domain_id(),
+							)
+						);
+						$urls[] = $url;
+
+						$domains[] = new Urlslab_Serp_Domain_Row(
+							array(
+								'domain_id'   => $url->get_domain_id(),
+								'domain_name' => $url_obj->get_domain_name(),
+							),
+							false
+						);
+
+						$positions[] = new Urlslab_Serp_Position_Row(
+							array(
+								'position' => $organic_result->position,
+								'query_id' => $query->get_query_id(),
+								'url_id'   => $url->get_url_id(),
+								'domain_id'   => $url->get_domain_id(),
+							)
+						);
+					}
+				}
+			}
+
+
+			if ( $has_monitored_domain ) {
+				if ( ! empty( $urls ) ) {
+					$urls[0]->insert_all( $urls, true );
+				}
+				if ( ! empty( $positions ) ) {
+					$positions[0]->insert_all( $positions, true );
+				}
+				if ( ! empty( $domains ) ) {
+					$domains[0]->insert_all( $domains, true );
+				}
+
+				$fqs = $serp_response->getFaqs();
+				if ( ! empty( $fqs ) && $widget->get_option( Urlslab_Serp::SETTING_NAME_IMPORT_FAQS ) ) {
+					foreach ( $fqs as $faq ) {
+						$faq_row = new Urlslab_Faq_Row( array( 'question' => $faq->question ), false );
+						if ( $faq_row->load( array( 'question' ) ) ) {
+							continue;
+						}
+						$faq_row->set_status( Urlslab_Faq_Row::STATUS_EMPTY );
+						$faq_row->insert();
+					}
+				}
+
+				$related = $serp_response->getRelatedSearches();
+				if ( ! empty( $related ) && $widget->get_option( Urlslab_Serp::SETTING_NAME_IMPORT_RELATED_QUERIES ) && $this->get_serp_queries_count() <= $widget->get_option( Urlslab_Serp::SETTING_NAME_IMPORT_LIMIT ) ) {
```

Comment:
shouldn't we check the limit before processing a keyword? this will ensure exact crawl